File: android/app/src/main/java/com/nauhaan/skycast/data/repository/LocationRepositoryImpl.kt
package com.nauhaan.skycast.data.repository

import com.nauhaan.skycast.core.common.DispatcherProvider
import com.nauhaan.skycast.data.local.dao.SavedLocationDao
import com.nauhaan.skycast.data.mapper.WeatherMapper.toDomain
import com.nauhaan.skycast.data.mapper.WeatherMapper.toEntity
import com.nauhaan.skycast.data.remote.ErrorMapper
import com.nauhaan.skycast.data.remote.OpenWeatherApi
import com.nauhaan.skycast.domain.model.LocationSearchResult
import com.nauhaan.skycast.domain.model.SavedLocation
import com.nauhaan.skycast.domain.repository.LocationRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Saved locations (Room) plus geocoding search (network).
 *
 * Unlike weather, search results are **not** cached: a place search is cheap, its
 * results are transient, and caching them would show the user stale matches for a
 * query they have since changed.
 */
@Singleton
class LocationRepositoryImpl
@Inject
constructor(
    private val dao: SavedLocationDao,
    private val api: OpenWeatherApi,
    private val dispatchers: DispatcherProvider,
) : LocationRepository {
    override fun observeSavedLocations(): Flow<List<SavedLocation>> =
        dao.observeAll().map { entities -> entities.map { it.toDomain() } }

    override fun observePrimaryLocation(): Flow<SavedLocation?> = dao.observePrimary().map { it?.toDomain() }

    override suspend fun getById(id: Long): SavedLocation? = withContext(dispatchers.io) {
        dao.getById(id)?.toDomain()
    }

    override suspend fun search(query: String): List<LocationSearchResult> = withContext(dispatchers.io) {
        val trimmed = query.trim()
        // Guard here rather than in the view model so every caller is protected
        // from burning API quota on an empty or one-character query.
        if (trimmed.length < MIN_QUERY_LENGTH) return@withContext emptyList()

        try {
            api.searchLocations(trimmed).map { it.toDomain() }
        } catch (throwable: Throwable) {
            throw ErrorMapper.map(throwable)
        }
    }

    override suspend fun save(result: LocationSearchResult): Long = withContext(dispatchers.io) {
        try {
            // The very first location the user adds becomes primary, so the Today tab
            // is never left with nothing to show.
            val isFirst = dao.count() == 0
            val id = dao.insert(result.toEntity(sortOrder = dao.nextSortOrder(), isPrimary = isFirst))
            if (isFirst) dao.setPrimaryExclusively(id)
            id
        } catch (throwable: Throwable) {
            throw ErrorMapper.map(throwable)
        }
    }

    override suspend fun delete(location: SavedLocation) = withContext(dispatchers.io) {
        // Cached weather rows cascade automatically via the entity's foreign key.
        dao.deleteAndReassignPrimary(location.toEntity())
    }

    override suspend fun setPrimary(location: SavedLocation) = withContext(dispatchers.io) {
        dao.setPrimaryExclusively(location.id)
    }

    override suspend fun reorder(orderedIds: List<Long>) = withContext(dispatchers.io) {
        dao.applyOrder(orderedIds)
    }

    private companion object {
        /** Below this length OpenWeather's geocoder returns noise, not matches. */
        const val MIN_QUERY_LENGTH = 2
    }
}

File: ios/SkyCast/Data/Repository/LocationRepositoryImpl.swift
import Foundation

/// Saved locations (SwiftData) plus geocoding search (network).
///
/// Search results are **not** cached, so a changed query cannot show stale matches.
final class LocationRepositoryImpl: LocationRepository {
    private let api: any WeatherAPI
    private let local: LocalDataStore

    init(api: any WeatherAPI, local: LocalDataStore) {
        self.api = api
        self.local = local
    }

    func savedLocations() async throws -> [SavedLocation] {
        do {
            return try await local.savedLocations()
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func primaryLocation() async throws -> SavedLocation? {
        do {
            return try await local.primaryLocation()
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func location(id: Int64) async throws -> SavedLocation? {
        do {
            return try await local.location(id: id)
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func search(query: String) async throws -> [LocationSearchResult] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        // Guarded here rather than in the view model so every caller is protected from
        // burning API quota on an empty or one-character query.
        guard trimmed.count >= Constants.minimumQueryLength else { return [] }

        let dtos = try await api.searchLocations(
            query: trimmed,
            limit: Constants.searchLimit
        )
        return dtos.map(WeatherMapper.searchResult(from:))
    }

    @discardableResult
    func save(_ result: LocationSearchResult) async throws -> Int64 {
        // The cap is enforced here rather than in the UI, because the UI is not the only caller and "the
        // list is full" is a fact about the data, not about a screen. See `SavedLocation.maxSaved`.
        let saved = try await savedLocations()
        guard SavedLocation.canSaveAnother(currentCount: saved.count) else {
            throw AppError.locationLimitReached(limit: SavedLocation.maxSaved)
        }

        do {
            return try await local.save(result)
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func delete(_ location: SavedLocation) async throws {
        do {
            try await local.delete(id: location.id)
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func setPrimary(_ location: SavedLocation) async throws {
        do {
            try await local.setPrimary(id: location.id)
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    func reorder(ids: [Int64]) async throws {
        do {
            try await local.reorder(ids: ids)
        } catch {
            throw AppError.storage(detail: error.localizedDescription)
        }
    }

    private enum Constants {
        /// Below this length OpenWeather's geocoder returns noise, not matches.
        static let minimumQueryLength = 2
        static let searchLimit = 8
    }
}
